cpp/server/proxy.h

```diff
@@ -0,0 +1,170 @@
+#ifndef CERT_TRANS_SERVER_PROXY_H_
+#define CERT_TRANS_SERVER_PROXY_H_
+
+#include <event2/http.h>
+#include <evhttp.h>
+#include <functional>
+#include <vector>
+
+#include "base/macros.h"
+#include "log/cluster_state_controller.h"
+#include "monitoring/monitoring.h"
+#include "net/url_fetcher.h"
+
+
+namespace util {
+class Task;
+}  // namespace util
+
+
+namespace cert_trans {
+
+namespace {
+static Counter<std::string>* total_proxied_requests(
+    Counter<std::string>::New("total_proxied_requests", "path",
+                              "Number of proxied API requests by path."));
+static Counter<std::string, int>* total_proxied_responses(
+    Counter<std::string, int>::New("total_proxied_responses", "path",
+                                   "status_code",
+                                   "Number of proxied API requests by path "
+                                   "and status code."));
+}  // namespace
+
+
+template <class Logged>
+class ClusterStateController;
+class UrlFetcher;
+
+
+template <class Logged>
```

Comment:
I'm not sure I'm entirely behind the de-templating idea; quite a lot of work has gone in to making the code be [relatively] *thing* agnostic (despite there only being a `LoggedCertificate` *thing* currently), and as you know there's interest in things like binary transparency, key transparency, etc. and I know of at least one currently active project who are interested in re-using this codebase for non-`LoggedCertificate` *things*.

That said, there is a lot of code here which is templated to actual *thing* where it really only needs access to a common `LoggedEntry` interface, so there's still a project for a rainy day :)